### PR TITLE
Fix iOS embedder memory management and other analyzer warnings

### PR DESCRIPTION
**Pull request by jmagman**
Run the static analyzer on the iOS embedder. Fix the issues, mostly related to memory management.

Hopefully soon we can run clang-tidy in post-submit or the push hooks https://github.com/flutter/flutter/issues/91857

To run clang-tidy for iOS:
```
dart tools/clang_tidy/bin/main.dart --lint-all --compile-commands out/ios_debug/compile_commands.json --repo .
```

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Headers/FlutterCallbackCache.h [hunk 1]
```diff
@@ -18,15 +18,15 @@ FLUTTER_DARWIN_EXPORT
 /**
  * The name of the callback.
  */
-@property(retain) NSString* callbackName;
+@property(copy) NSString* callbackName;
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Headers/FlutterCallbackCache.h:21]: Foundation objects with mutable variants (`NSMutableString`) should be `copy`.

File: shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm [hunk 2]
```diff
@@ -288,7 +288,8 @@ + (NSString*)domainNetworkPolicy:(NSDictionary*)appTransportSecurity {
   NSData* jsonData = [NSJSONSerialization dataWithJSONObject:networkConfigArray
                                                      options:0
                                                        error:NULL];
-  return [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
+  [networkConfigArray release];
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm:291]: `networkConfigArray` was leaked.

File: shell/platform/darwin/ios/framework/Source/FlutterEmbedderKeyResponder.mm [hunk 3]
```diff
@@ -349,15 +349,15 @@ @interface FlutterEmbedderKeyResponder ()
  *
  * Set by the initializer.
  */
-@property(nonatomic) FlutterSendKeyEvent sendEvent;
+@property(nonatomic, copy, readonly) FlutterSendKeyEvent sendEvent;
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterEmbedderKeyResponder.mm:352]: https://github.com/flutter/engine/blob/b7459d5058f6bfd7dbe41a7b92d46585cafab970/shell/platform/darwin/ios/framework/Source/FlutterEmbedderKeyResponder.mm#L487

File: shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm [hunk 4]
```diff
@@ -212,6 +212,7 @@ - (instancetype)initWithEnableObservatoryPublication:(BOOL)enableObservatoryPubl
               if (weak.get().enableObservatoryPublication) {
                 [[weak.get() delegate] publishServiceProtocolPort:url];
               }
+              [url release];
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterObservatoryPublisher.mm:215]: `url` was leaked.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 5]
```diff
@@ -671,21 +671,27 @@ - (void)viewDidLoad {
...
-  [self.keyboardManager release];
-  self.keyboardManager = [[FlutterKeyboardManager alloc] init];
+  FlutterKeyboardManager* keyboardManager = [[FlutterKeyboardManager alloc] init];
+  self.keyboardManager = keyboardManager;
+  [keyboardManager release];
+  fml::WeakPtr<FlutterViewController> weakSelf = [self getWeakPtr];
   FlutterSendKeyEvent sendEvent =
       ^(const FlutterKeyEvent& event, FlutterKeyEventCallback callback, void* userData) {
-        [_engine.get() sendKeyEvent:event callback:callback userData:userData];
+        [weakSelf.get()->_engine.get() sendKeyEvent:event callback:callback userData:userData];
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:680]: This didn't show up in the linter, but this retain cycle was exposed when I wrote the `testReleasesKeyboardManagerOnDealloc` and the view controller wasn't being released.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 6]
```diff
@@ -1729,6 +1735,7 @@ - (void)scrollEvent:(UIPanGestureRecognizer*)recognizer API_AVAILABLE(ios(13.4))
 - (void)encodeRestorableStateWithCoder:(NSCoder*)coder {
   NSData* restorationData = [[_engine.get() restorationPlugin] restorationData];
   [coder encodeDataObject:restorationData];
+  [super encodeRestorableStateWithCoder:coder];
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:1738]: This was an analyzer warning, `encodeRestorableStateWithCoder` needs to encode the super class's state.

File: shell/platform/darwin/ios/rendering_api_selection.mm [hunk 7]
```diff
@@ -22,6 +22,7 @@ bool ShouldUseMetalRenderer() {
   if (@available(iOS METAL_IOS_VERSION_BASELINE, *)) {
     auto device = MTLCreateSystemDefaultDevice();
     ios_version_supports_metal = [device supportsFeatureSet:MTLFeatureSet_iOS_GPUFamily1_v3];
+    [device release];
```
Comment (jmagman) [inline shell/platform/darwin/ios/rendering_api_selection.mm:25]: `MTLCreateSystemDefaultDevice` is `NS_RETURNS_RETAINED`

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 8]
```diff
@@ -671,21 +671,27 @@ - (void)viewDidLoad {
...
       addPrimaryResponder:[[FlutterEmbedderKeyResponder alloc] initWithSendEvent:sendEvent]];
-  [self.keyboardManager addPrimaryResponder:[[FlutterChannelKeyResponder alloc]
-                                                initWithChannel:self.engine.keyEventChannel]];
-  [self.keyboardManager addSecondaryResponder:self.engine.textInputPlugin];
+  FlutterChannelKeyResponder* responder =
+      [[FlutterChannelKeyResponder alloc] initWithChannel:self.engine.keyEventChannel];
+  [self.keyboardManager addPrimaryResponder:responder];
+  [responder release];
+  FlutterTextInputPlugin* textInputPlugin = self.engine.textInputPlugin;
+  if (textInputPlugin != nil) {
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:689]: `self.engine` is nullable. When I tested it threw when adding nil to the secondary responder array.

File: shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm (hunk 2 above)
Comment (stuartmorgan) [reply shell/platform/darwin/ios/framework/Source/FlutterDartProject.mm:291]: The style in projects I worked on prior to the advent of ARC was that unless there was a compelling reason not to--e.g., code that's called in a tight loop--it's far better to `autorelease` as part of the `alloc/init` statement, because then there's no chance of introducing a leak later via early return. It also made it much easier to review code for correctness, because any line that did an `alloc/init` but not an `autorelease` was immediately suspect, triggering a check for a corresponding `release`.

So tl;dr, I would make this part of line 276 instead.

File: shell/platform/darwin/ios/framework/Source/FlutterEmbedderKeyResponder.mm [hunk 9]
```diff
@@ -349,15 +349,15 @@ @interface FlutterEmbedderKeyResponder ()
...
 
 /**
  * A map of pressed keys.
  *
  * The keys of the dictionary are physical keys, while the values are the logical keys
  * of the key down event.
  */
-@property(nonatomic) NSMutableDictionary<NSNumber*, NSNumber*>* pressingRecords;
+@property(nonatomic, retain, readonly) NSMutableDictionary<NSNumber*, NSNumber*>* pressingRecords;
```
Comment (stuartmorgan) [inline shell/platform/darwin/ios/framework/Source/FlutterEmbedderKeyResponder.mm:360]: Why do we have `readonly` private properties?

File: shell/platform/darwin/ios/framework/Source/FlutterKeyboardManager.mm [hunk 10]
```diff
@@ -13,12 +13,14 @@ @interface FlutterKeyboardManager ()
 /**
  * The primary responders added by addPrimaryResponder.
  */
-@property(nonatomic) NSMutableArray<id<FlutterKeyPrimaryResponder>>* primaryResponders;
+@property(nonatomic, retain, readonly)
```
Comment (stuartmorgan) [inline shell/platform/darwin/ios/framework/Source/FlutterKeyboardManager.mm:16]: Same question here.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 11]
```diff
@@ -671,21 +671,27 @@ - (void)viewDidLoad {
 }
 
 - (void)addInternalPlugins {
-  [self.keyboardManager release];
-  self.keyboardManager = [[FlutterKeyboardManager alloc] init];
+  FlutterKeyboardManager* keyboardManager = [[FlutterKeyboardManager alloc] init];
```
Comment (stuartmorgan) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:674]: Why is this local being introduced?

```
self.keyboard = [[[FlutterKeyboardManager alloc] init] autorelease];
```

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 12]
```diff
@@ -671,21 +671,27 @@ - (void)viewDidLoad {
...
       };
   [self.keyboardManager
       addPrimaryResponder:[[FlutterEmbedderKeyResponder alloc] initWithSendEvent:sendEvent]];
-  [self.keyboardManager addPrimaryResponder:[[FlutterChannelKeyResponder alloc]
-                                                initWithChannel:self.engine.keyEventChannel]];
-  [self.keyboardManager addSecondaryResponder:self.engine.textInputPlugin];
+  FlutterChannelKeyResponder* responder =
+      [[FlutterChannelKeyResponder alloc] initWithChannel:self.engine.keyEventChannel];
+  [self.keyboardManager addPrimaryResponder:responder];
+  [responder release];
```
Comment (stuartmorgan) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:687]: Same here; the autorelease can be inlined, removing the need for a local.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController.mm [hunk 13]
```diff
@@ -671,21 +671,27 @@ - (void)viewDidLoad {
...
+  FlutterTextInputPlugin* textInputPlugin = self.engine.textInputPlugin;
+  if (textInputPlugin != nil) {
+    [self.keyboardManager addSecondaryResponder:self.engine.textInputPlugin];
```
Comment (stuartmorgan) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController.mm:690]: s/self.engine//

File: shell/platform/darwin/ios/framework/Source/FlutterViewController_Internal.h [hunk 14]
```diff
@@ -31,7 +31,7 @@ extern NSNotificationName const FlutterViewControllerShowHomeIndicator;
 
 @property(nonatomic, readonly) BOOL isPresentingViewController;
 @property(nonatomic, readonly) BOOL isVoiceOverRunning;
-@property(nonatomic) FlutterKeyboardManager* keyboardManager;
+@property(nonatomic, strong) FlutterKeyboardManager* keyboardManager;
```
Comment (stuartmorgan) [inline shell/platform/darwin/ios/framework/Source/FlutterViewController_Internal.h:34]: Isn't `strong` an ARC annotation?

File: shell/platform/darwin/ios/framework/Source/SemanticsObject.mm [hunk 15]
```diff
@@ -433,7 +433,7 @@ - (NSAttributedString*)createAttributedStringFromString:(NSString*)string
       }
     }
   }
-  return attributedString;
+  return [attributedString autorelease];
```
Comment (stuartmorgan) [inline shell/platform/darwin/ios/framework/Source/SemanticsObject.mm:436]: This should be part of line 412.

**Review by stuartmorgan — APPROVED** [review]
LGTM with nits

File: shell/platform/darwin/ios/framework/Source/FlutterViewController_Internal.h (hunk 14 above)
Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController_Internal.h:34]: Yes, whoops.

File: shell/platform/darwin/ios/framework/Source/FlutterUmbrellaImport.m [hunk 16]
```diff
@@ -2,6 +2,8 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
 
+// FLUTTER_NOLINT: https://github.com/flutter/flutter/issues/93360
```
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/FlutterUmbrellaImport.m:5]: This one is new.
[`FlutterUmbrellaImport`](https://github.com/flutter/engine/blob/cf087516afb930de549e672b37a1e7d70b379931/shell/platform/darwin/ios/framework/Source/FlutterUmbrellaImport.m#L5-L9) is used in the [`copy_and_verify_framework_module` target](https://github.com/flutter/engine/blob/d57627c98fee5da1eb1a85ae29cbd083fbc98c5c/shell/platform/darwin/ios/BUILD.gn#L380), is not included in the engine framework source, and cannot be compiled on its own.

When `clang-tidy` is run on this file, it fails with a `clang-diagnostic-error` (compilation failure).  Add a `FLUTTER_NOLINT` to this file with link to https://github.com/flutter/flutter/issues/93360.

File: shell/platform/darwin/ios/framework/Source/FlutterViewController_Internal.h (hunk 14 above)
Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/FlutterViewController_Internal.h]: Interesting, switching all the `strong`s (outside the public headers) to `retain`s exposed a few more missing `dealloc`s.

File: shell/platform/darwin/ios/framework/Source/FlutterEmbedderKeyResponder.mm (hunk 9 above)
Comment (stuartmorgan) [reply shell/platform/darwin/ios/framework/Source/FlutterEmbedderKeyResponder.mm:360]: I'm aware of the property vs ivar debate; I actually don't have a strong opinion on that. I was just curious about the addition of `readonly` to a private proprety. It just seems odd to say that code within a class can modify the collection, but not replace it with a different collection.

But this wasn't intended to be a PR-blocking question.

Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/FlutterEmbedderKeyResponder.mm:360]: > I'm aware of the property vs ivar debate; I actually don't have a strong opinion on that.
Guess I'm always geared up for a religious war in these kinds of reviews 😉 

Re: the readonly, you're right, that wasn't a static analyzer warning.  I did it because the property wasn't being set other than in `init`, so it could be made readonly, which I did in the spirit of preferring stricter property attributes, readonly, immutable, nonnull, etc.  But yeah, it's not API just private so it doesn't really matter (unless you're KVOing on the property and need to handle updating notifications in the property setter, which we aren't).  I can change it back if you prefer.
